### PR TITLE
PR: Add testing requirements to setup.py

### PR DESCRIPTION
**Pull request by rlaverde**
Fixes: #4174

-------
I take the packages that I found in the continuum integration recipes.

I'm not sure if `numpy` should also be added (pandas will install it as a dependency), and if `pillow` if needed (circle-ci recipe doesn't install it)

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -288,8 +288,15 @@ def run(self):
     'numpydoc',
 ]
 
+extra_require = {
```
Comment (ccordoba12) [inline setup.py]: `extra_require` -> `extras_require`

File: setup.py [hunk 2]
```diff
@@ -288,8 +288,15 @@ def run(self):
...
+    'test:python_version == "2.7"': ['mock'],
```
Comment (ccordoba12) [inline setup.py]: How does this syntax work? I'm just curious and I'd like to understand how it works :-)

File: setup.py (hunk 1 above)
Comment (ccordoba12) [inline setup.py]: The failure in Travis is caused because of this.

**Comment by ccordoba12** [comment]
> I'm not sure if numpy should also be added (pandas will install it as a dependency), and if pillow if needed (circle-ci recipe doesn't install it)

Ok, no problem about Numpy., i.e. you can remove it.

Pillow is used to generate image representations of arrays in the Variable Explorer. So please add a test for that and leave it as a test dependency ;-)

File: setup.py (hunk 2 above)
Comment (rlaverde) [inline setup.py]: I only found that documented in the[ wheels docs](https://wheel.readthedocs.io/en/latest/#defining-conditional-dependencies)

The idea is that you could run:
`pip install spyder[test]`
and It will install this extra dependencies

**Comment by ccordoba12** [comment]
Ok, good. Now please modify our Travis scripts to install our wheel with `[test]`, then install `pyqt5` (i.e. remove our conda install of pyqt5 for pip) and finally run our pytest's there.

The idea is to use our Travis pip slot to test the latest PyQt 5 (which is 5.8 right now) because conda is going to stay at 5.6 for a loong time ;-)

File: setup.py [hunk 3]
```diff
@@ -288,8 +288,15 @@ def run(self):
     'numpydoc',
 ]
 
+extras_require = {
+    'test:python_version == "2.7"': ['mock'],
+    'test': ['pytest', 'pytest-qt', 'pytest-cov', 'mock', 'coveralls',
+             'ciocheck', 'flaky', 'pandas', 'scipy', 'sympy', 'pillow'],
```
Comment (ccordoba12) [inline setup.py]: `ciocheck` is not in PiPy. Please remove it.

File: continuous_integration/travis/install.sh [hunk 4]
```diff
@@ -18,6 +18,7 @@ download_code()
     if [ "$PR" != "false" ] ; then
         cd $FULL_SPYDER_CLONE
         git fetch origin pull/$PR/head:travis_pr_$PR
+        git checkout travis_pr_$TRAVIS_PULL_REQUEST
```
Comment (ccordoba12) [inline continuous_integration/travis/install.sh]: This is done in `build_test.sh`, so you don't need to do it here:

https://github.com/spyder-ide/spyder/blob/master/continuous_integration/travis/build_test.sh#L10

File: setup.py (hunk 3 above)
Comment (rlaverde) [inline setup.py]: but it is listed in pipy https://pypi.python.org/pypi/ciocheck/0.1.1

File: continuous_integration/travis/install.sh [hunk 5]
```diff
@@ -61,14 +62,10 @@ install_conda()
...
-    conda install 'pyqt>=5.6.0';
+    pip install pyqt5;
 
     # Install testing packages
-    pip install pytest pytest-cov pytest-qt mock
-
-    # Install extra packages
-    EXTRA_PACKAGES="matplotlib pandas sympy pyzmq pillow"
-    pip install $EXTRA_PACKAGES
+    pip install -e .[test]
```
Comment (ccordoba12) [inline continuous_integration/travis/install.sh]: This is not needed here. Please do it after the wheel has been built, i.e. here:

https://github.com/spyder-ide/spyder/blob/master/continuous_integration/travis/build_test.sh#L10

Comment (rlaverde) [inline continuous_integration/travis/install.sh]: ok, this will allow to delete the duplicated checkout

File: continuous_integration/travis/install.sh (hunk 4 above)
Comment (rlaverde) [inline continuous_integration/travis/install.sh]: yes, but if I try to run `pip install -e .[test]` It'll fail because It was in the master branch

File: setup.py (hunk 3 above)
Comment (ccordoba12) [inline setup.py]: But Travis is complaining about the lack of a package for ciocheck.

In any case, pytest doesn't run ciocheck, so it's not really a test dependency.

File: continuous_integration/travis/run_test.sh [hunk 6]
```diff
@@ -26,6 +26,9 @@ if [ "$USE_CONDA" = true ] ; then
 else
     cd $FULL_SPYDER_CLONE
     pip install dist/spyder-*.whl
+
+    # Install testing packages
+    pip install -e .[test]
```
Comment (ccordoba12) [inline continuous_integration/travis/run_test.sh]: Is this line necessary? I mean, what I want is to install our wheel and see if it pulls our test dependencies. So something like

     pip install dist/spyder-*.whl[test]

should work (but I'm not sure :-)

Comment (rlaverde) [inline continuous_integration/travis/run_test.sh]: I'm not sure neither :see_no_evil:

I added a commit to see if this work

Comment (rlaverde) [inline continuous_integration/travis/run_test.sh]: It didn't work because `dist/spyder-*.whl` is a filename

**Comment by rlaverde** [comment]
@ccordoba12 I think the installation using pip is partially working, but I'm blocked because the issue with QtWebEngineWidgets not avalaible in debian/ubuntu

File: continuous_integration/travis/install.sh [hunk 7]
```diff
@@ -61,14 +61,7 @@ install_conda()
...
 {
     # Install PyQt5
-    conda install 'pyqt>=5.6.0';
-
-    # Install testing packages
-    pip install pytest pytest-cov pytest-qt mock
-
-    # Install extra packages
-    EXTRA_PACKAGES="matplotlib pandas sympy pyzmq pillow"
-    pip install $EXTRA_PACKAGES
```
Comment (ccordoba12) [inline continuous_integration/travis/install.sh]: Please revert this change for now. I have to plan on how to start using the pyqt5 wheels on CircleCI.

File: continuous_integration/travis/run_test.sh (hunk 6 above)
Comment (ccordoba12) [inline continuous_integration/travis/run_test.sh]: I'll fix this one before merging this PR ;-)

File: continuous_integration/travis/install.sh (hunk 7 above)
Comment (rlaverde) [inline continuous_integration/travis/install.sh]: I only revert the installation of pyqt (to install it with conda), the other dependencies will be installed by `pip install -e .[test]`

Comment (ccordoba12) [inline continuous_integration/travis/install.sh]: You're totally right :-)

**Comment by ccordoba12** [comment]
@ghisvail, I think this could be useful for you guys if you want to run our tests in the Debian infrastructure.

**Comment by ghisvail** [comment]
Could be, assuming you guys distribute your tests following the pytest [guidelines](http://doc.pytest.org/en/latest/goodpractices.html#choosing-a-test-layout-import-rules). The Debian packaging is not currently CI tested on the Debian infrastructure.

Another thing which could help would be to verify whether `spyder` can be tested with `pytest-xvfb` instead of manual calls to `xvfb` / `xvfb-run`. The latter are known to be occasionally unreliable on parallel execution setups in our experience. Same comment applies to `qtpy` and `qtawesome`. It worked for `spyder-unittest`.

**Comment by ccordoba12** [comment]
> assuming you guys distribute your tests following the pytest guidelines

I think we are, by "inlining test directories into [our] application package"

> Another thing which could help would be to verify whether spyder can be tested with pytest-xvfb

CircleCi always starts a xvfb instance, so we haven't seen the need of using pytest-xvfb. However, I'm not opposed to idea :-) Could you open a PR to see how that could work?
